sor/src/main/java/com/bazaarvoice/emodb/sor/db/astyanax/CqlDataWriterDAO.java

```diff
@@ -0,0 +1,196 @@
+        for (AstyanaxStorage storage : table.getWriteStorage()) {
+            DeltaPlacement placement = (DeltaPlacement) storage.getPlacement();
+            CassandraKeyspace keyspace = placement.getKeyspace();
+
+            ByteBuffer rowKey = storage.getRowKey(key);
+
+            // Should synchronously write compaction and then delete deltas
+            writeCompaction(rowKey, compactionKey, compaction, consistency, placement, keyspace, tbl, key);
+
+            deleteCompactedDeltas(rowKey, consistency, placement, keyspace, changesToDelete, historyList, tbl, key);
+        }
+    }
+
+    private void insertBlockedDeltas(BatchStatement batchStatement, DeltaTableDDL tableDDL, ConsistencyLevel consistencyLevel, ByteBuffer rowKey, UUID changeId, ByteBuffer encodedDelta) {
+
+        List<ByteBuffer> blocks = DAOUtils.getBlockedDeltas(encodedDelta, _deltaPrefixLength, _deltaBlockSize);
+
+        for (int i = 0; i < blocks.size(); i++) {
+            batchStatement.add(QueryBuilder.insertInto(tableDDL.getTableMetadata())
+                    .value(tableDDL.getRowKeyColumnName(), rowKey)
+                    .value(tableDDL.getChangeIdColumnName(), changeId)
+                    .value(tableDDL.getBlockColumnName(), i)
+                    .value(tableDDL.getValueColumnName(), blocks.get(i))
+                    .setConsistencyLevel(consistencyLevel));
+        }
+    }
+
+    private void writeCompaction(ByteBuffer rowKey, UUID compactionKey, Compaction compaction,
+                                 WriteConsistency consistency, DeltaPlacement placement,
+                                 CassandraKeyspace keyspace, Table table, String key) {
+
+        // Add the compaction record
+        ByteBuffer encodedCompaction = ByteBuffer.wrap(_changeEncoder.encodeCompaction(compaction, new StringBuilder(_deltaPrefix)).toString().getBytes());
+
+        Session session = keyspace.getCqlSession();
+        ConsistencyLevel consistencyLevel = SorConsistencies.toCql(consistency);
+        DeltaTableDDL tableDDL = placement.getDeltaTableDDL();
+
+        BatchStatement batchStatement = new BatchStatement(BatchStatement.Type.LOGGED);
+
+        insertBlockedDeltas(batchStatement, tableDDL, consistencyLevel, rowKey, compactionKey, encodedCompaction);
+
+        session.execute(batchStatement);
+    }
+
+    private void deleteCompactedDeltas(ByteBuffer rowKey, WriteConsistency consistency, DeltaPlacement placement,
+                                       CassandraKeyspace keyspace, Collection<UUID> changesToDelete,
+                                       List<History> historyList, Table table, String key) {
+
+        Session session = keyspace.getCqlSession();
+        ConsistencyLevel consistencyLevel = SorConsistencies.toCql(consistency);
+
+        // delete the old deltas & compaction records
+
+        DeltaTableDDL tableDDL = placement.getDeltaTableDDL();
+
+        // each individual delete statement will be atomic, so the batch can be unlogged
```

Comment:
Maybe even include a comment such as "if any delete fails in this batch it will be deleted again the next time it is read"